Simplified/Audiobooks/NYPLAudiobookDownloader.swift

```diff
@@ -0,0 +1,142 @@
+//
+//  NYPLAudiobooksDownloader.swift
+//  Simplified
+//
+//  Created by Ernest Fan on 2022-04-01.
+//  Copyright © 2022 NYPL. All rights reserved.
+//
+#if FEATURE_AUDIOBOOKS
+import Foundation
+import NYPLAudiobookToolkit
+
+@objc protocol NYPLAudiobookDownloadStatusDelegate {
+  func audiobookDidUpdateDownloadProgress(progress: Float, bookID: String)
+  func audiobookDidCompleteDownload(bookID: String)
+  func audiobookDidCompleteDownloadFirstElement(bookID: String)
+  func audiobookDidReceiveDownloadError(error: NSError?, bookID: String)
+}
+
+class NYPLAudiobookDownloadObject {
+  var bookID: String
+  var audiobookManager: DefaultAudiobookManager
+  
+  init(bookID: String, audiobookManager: DefaultAudiobookManager) {
+    self.bookID = bookID
+    self.audiobookManager = audiobookManager
+  }
+}
+
+@objcMembers class NYPLAudiobookDownloader: NSObject {
+  weak var delegate: NYPLAudiobookDownloadStatusDelegate?
+  
+  private var serialQueue: DispatchQueue = DispatchQueue(label: "org.nypl.labs.NYPLAudiobooksDownloader")
+  private var downloadObjects: [NYPLAudiobookDownloadObject] = []
+  private var currentDownloadObject: NYPLAudiobookDownloadObject?
+  
+  @objc(downloadAudiobookForBookID:audiobookManager:)
+  func downloadAudiobook(for bookID: String,
+                         audiobookManager: DefaultAudiobookManager) {
+    guard downloadObject(for: bookID) == nil else {
+      return
+    }
+    
+    serialQueue.async {
+      self.downloadObjects.append(NYPLAudiobookDownloadObject(bookID: bookID,
+                                                              audiobookManager: audiobookManager))
+    }
+    fetchNext()
+  }
+  
+  @objc func cancelDownload(for bookID: String) {
+    if let downloadObject = currentDownloadObject,
+       downloadObject.bookID == bookID {
+      cancelDownload(for: downloadObject)
+      releaseCurrentDownloadObject()
+      fetchNext()
```

Comment:
When the current audiobook download is cancelled, we will start fetching the next audiobook in queue. Maybe it should be called `fetchNextAudiobookInQueue`?

Reply:
I'd rather rename `cancelDownload(for:)` to make it explicit that we're trying to fetch the next too, like "cancelAndFetchNext" or something. But ideally it would be better if we can more the "fetch" part outside of the "cancel" method, unless that complicates things on the caller side significantly. Generally speaking, it's better to have more focused methods than methods that do more than one thing.

Reply:
I didn't put two methods together because user is allowed to cancel downloads that has not started, aka still in the queue waiting. For such case, we do not want to fetch the next audiobook in queue.